cpp/src/arrow/dataset/partition.h

```diff
@@ -18,87 +18,114 @@
 /// store instead of in file paths, for example
 /// dataset_root/2009/11/... could be used when the partition fields
 /// are "year" and "month"
-struct PartitionKeyData {
-  std::vector<std::string> fields;
-  std::vector<std::shared_ptr<Scalar>> values;
-};
-
-// ----------------------------------------------------------------------
-// Partition schemes
-
-/// \brief
+///
+/// Paths are consumed from left to right. Paths must be relative to
+/// the root of a partition; path prefixes must be removed before passing
+/// the path to a scheme for parsing.
 class ARROW_DS_EXPORT PartitionScheme {
  public:
   virtual ~PartitionScheme() = default;
 
   /// \brief The name identifying the kind of partition scheme
   virtual std::string name() const = 0;
 
-  virtual bool PathMatchesScheme(const std::string& path) const = 0;
+  /// \brief Parse a path into a partition expression
+  ///
+  /// \param[in] path the partition identifier to parse
+  /// \param[out] unconsumed a suffix of path which was not consumed
+  /// \param[out] out the parsed expression
+  virtual Status Parse(const std::string& path, std::string* unconsumed,
+                       std::shared_ptr<Expression>* out) const = 0;
+
+  /// \brief Parse a path into a partition expression
+  ///
+  /// Helper for use by a root PartitionScheme; consumes whole path.
+  Status Parse(const std::string& path, std::shared_ptr<Expression>* out) const;
+};
+
+/// \brief Trivial partition scheme which only consumes a specified prefix (empty by
+/// default) from a path, yielding an expression provided on construction.
+class ARROW_DS_EXPORT SimplePartitionScheme : public PartitionScheme {
+ public:
+  explicit SimplePartitionScheme(std::shared_ptr<Expression> expr,
+                                 std::string ignored = "")
+      : partition_expression_(std::move(expr)), ignored_(std::move(ignored)) {}
+
+  std::string name() const override { return "simple_partition_scheme"; }
+
+  Status Parse(const std::string& path, std::string* unconsumed,
+               std::shared_ptr<Expression>* out) const override;
+
+ private:
+  std::shared_ptr<Expression> partition_expression_;
+  std::string ignored_;
+};
+
+/// \brief Combine partition schemes
+class ARROW_DS_EXPORT ChainPartitionScheme : public PartitionScheme {
+ public:
+  explicit ChainPartitionScheme(std::vector<std::unique_ptr<PartitionScheme>> schemes)
+      : schemes_(std::move(schemes)) {}
+
+  std::string name() const override { return "chain_partition_scheme"; }
+
+  Status Parse(const std::string& path, std::string* unconsumed,
+               std::shared_ptr<Expression>* out) const override;
+
+ protected:
+  std::vector<std::unique_ptr<PartitionScheme>> schemes_;
+};
+
+/// \brief Parse a single field from a single path segment
+class ARROW_DS_EXPORT FieldPartitionScheme : public PartitionScheme {
+ public:
+  explicit FieldPartitionScheme(std::shared_ptr<Field> field)
```

Comment:
I know chain exists, but why not also take a schema and consume paths?

```f = FieldPartitionScheme(schema({field("year", int64()), field("month", int64()), field("day", int64())}))
f.Parse("/2017") = { "year": 2017 }
f.Parse("/2017/05") = { "year": 2017, "month": 05 }
f.Parse("/2017/05/12") = { "year": 2017, "month": 05, "day": 12}
```

Reply:
Since the field names are not present in the path, that would rely on ordering of fields within a schema (which is supposed to be an associative structure)

Reply:
Correct. But the schema is ordered (as per the Equals method).

Reply:
renamed to `SchemaPartitionScheme`